NMS-13362: Add Kafka consumer for OpenNMS events

cgorantla opened the pull request:
Add new  feature `opennms-kafka-consumer` that allows to consume events from Kafka.
This allows OpenNMS to consume events that are generated when OpenNMS is Offline.

Also it simplifies the event generation to be protobuf.

### External References

* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-13362

cgorantla commented on features/kafka/consumer/src/main/proto/kafka-consumer-events.proto:
```diff
@@ -0,0 +1,37 @@
+    string value = 2;
+}
+
+message Event {
```
@j-white @RangerRick   I brought most of the elements in `org.opennms.netmgt.xml.event.Event` to proto.
Let me know if this looks fine.

cgorantla commented:
Currently this is targeting to `develop` but think this can be backported to `foundation-2020`

indigo423 commented on docs/modules/operation/pages/events/sources/kafka.adoc:
```diff
@@ -0,0 +1,50 @@
+The Kafka Consumer feature allows events to be consumed from specific topic on Kafka.
+
+Events should be created in protobuf format link:https://developers.google.com/protocol-buffers/[Google Protocol Buffers (GPB)]
+See `kafka-consumer-events.proto`  for the model definitions.
```
```suggestion
See `kafka-consumer-events.proto` for the model definitions.
```

RangerRick commented on features/kafka/consumer/src/main/proto/kafka-consumer-events.proto:
```diff
@@ -0,0 +1,37 @@
+option java_outer_classname = "EventsProto";
+
+// The values differ from the standard codes in OpenNMS
+// since proto3 enforces us to start at 0
```
Oof... I don't suppose we could just make 0 be "unknown" or something, and have it not be used? kinda sucks to have them off-by-one.

Naicisum replied:
Why not do the industry standard, use `UNUSED`, `PAD0`, `PAD1`, or `RESERVED`?

cgorantla replied:
`INDETERMINATE` seems to be 0 so there is already `unknown` case.  Mapping seems to be easier by matching label.

cgorantla replied:
@Naicisum I think this issue is mostly trying to use our `Event` and match it with protobuf and get events consumed directly on OpenNMS.

RangerRick replied:
Sure, but `INDETERMINATE` has a specific meaning/mapping in the OpenNMS code, whereas in this case the 0th entry would be not actually used.  I am not a fan of having 2 different "standards" for what the enums map top (0-indexed and 1-indexed), it makes it the API ambiguous and have different behavior whether you're parsing say, the ReST API vs the Kafka stream.

j-white replied:
This was copied from here [opennms-kafka-producer.proto#L7](https://github.com/OpenNMS/opennms/blob/opennms-28.0.0-2/features/kafka/producer/src/main/proto/opennms-kafka-producer.proto#L7) for reference - so we already to the start at 0 thing in the producer case.

RangerRick replied:
oof, alright... then I guess there are [2 standards](https://xkcd.com/927/)  :D

indigo423 commented on docs/modules/operation/pages/events/sources/kafka.adoc:
```diff
@@ -0,0 +1,50 @@
+
+[source]
+----
+$ ssh -p 8101 admin@localhost
```
```suggestion
ssh -p 8101 admin@localhost
```

indigo423 replied:
We should split the source sections to indicate what is console and what is Karaf shell like this:

```
.Configure the Kafka features and the Minion location via the Karaf Shell
[source, console]
----
ssh -p 8201 admin@localhost
----

.Configure the Event consumer client
[source, karaf]
----
config:edit org.opennms.features.kafka.consumer.client
config:property-set bootstrap.servers 127.0.0.1:9092
config:update
----
```

indigo423 commented on docs/modules/operation/pages/events/sources/kafka.adoc:
```diff
@@ -0,0 +1,50 @@
+----
+$ ssh -p 8101 admin@localhost
+...
+admin@opennms()> config:edit org.opennms.features.kafka.consumer.client
```
```suggestion
config:edit org.opennms.features.kafka.consumer.client
```

indigo423 commented on docs/modules/operation/pages/events/sources/kafka.adoc:
```diff
@@ -0,0 +1,50 @@
+$ ssh -p 8101 admin@localhost
+...
+admin@opennms()> config:edit org.opennms.features.kafka.consumer.client
+admin@opennms()> config:property-set bootstrap.servers 127.0.0.1:9092
```
```suggestion
config:property-set bootstrap.servers 127.0.0.1:9092
```

indigo423 replied:
We should probably add a callout for the 127.0.0.1 bootstrap server and port like this:

> Connect to the following Kafka nodes and adjust the IPs or FQDNs with the Kafka port (9092) accordingly.

indigo423 commented on docs/modules/operation/pages/events/sources/kafka.adoc:
```diff
@@ -0,0 +1,50 @@
+...
+admin@opennms()> config:edit org.opennms.features.kafka.consumer.client
+admin@opennms()> config:property-set bootstrap.servers 127.0.0.1:9092
+admin@opennms()> config:update
```
```suggestion
config:update
```

indigo423 commented on docs/modules/operation/pages/events/sources/kafka.adoc:
```diff
@@ -0,0 +1,50 @@
+
+[source]
+----
+admin@opennms()> feature:install opennms-kafka-consumer
```
```suggestion
feature:install opennms-kafka-consumer
```

indigo423 commented on docs/modules/operation/pages/events/sources/kafka.adoc:
```diff
@@ -0,0 +1,50 @@
+First, log in to the Karaf shell of your {page-component-title} instance and configure the Kafka client settings to point to your Kafka broker.
+See link:https://kafka.apache.org/27/documentation.html#consumerconfigs[Consumer Configs] for a complete list of available options.
+
+[source]
```
```suggestion
[source, console]
```

indigo423 commented on docs/modules/operation/pages/events/sources/kafka.adoc:
```diff
@@ -0,0 +1,50 @@
+== Configure Topic Name
+
+.Karaf login, configuration of events topic
+[source]
```
Same as above, split and no prompts so we can use copy & paste.

indigo423 reviewed (changes requested):
Just minor things so terminal in and outputs are consistent.

Naicisum commented:
Probably need to code a runtime config check to ensure a loop between the kafka-producer event topic if configured, and this feature's event topic does not occur.

opennms-bamboo commented:
Welcome CI/CD System (cicd-system@opennms.com).

We would like to thank you for your contribution.

I could not find a signed OpenNMS Contributor Agreement (OCA) assigned to the username CI/CD System (cicd-system@opennms.com).
In order to maintain some order with respect to the copyright to the code,
every contributor is asked to sign a "Contributor Agreement".

The OpenNMS Contributor Agreement can be downloaded [here](https://wiki.opennms.org/wiki/images/6/6b/ContributorAgreement.pdf).
Please send a signed copy to oca@opennms.org.

If you need more information, please visit https://wiki.opennms.org/wiki/Contributor_Agreement

cgorantla commented:
> Probably need to code a runtime config check to ensure a loop between the kafka-producer event topic if configured, and this feature's event topic does not occur.

That's good point. We may need to check if topic currently exist before subscribing and we have to do that in lot of places. For now I renamed it as `opennms-kafka-events` and added a note on doc.

cgorantla commented:
@indigo423 Thanks for the review. Handled them. Please take a look again.

Bonrob2 commented on docs/modules/operation/pages/events/sources/kafka.adoc:
```diff
@@ -59,3 +59,4 @@ config:property-set eventsTopic opennms-kafka-events
 config:update
 ----
 
+NOTE: While configuring events topic, make sure that it doesn't conflict with other topics in OpenNMS subsystem. If you are unsure, keep the default one.
```
NOTE: While configuring the eventsTopic, make sure that it doesn't conflict with other topics in the OpenNMS subsystem. If you are unsure, keep the default one.

Bonrob2 replied:
```suggestion
NOTE: While configuring the eventsTopic, make sure that it doesn't conflict with other topics in the OpenNMS subsystem. If you are unsure, keep the default one.
```

Bonrob2 commented on docs/modules/operation/pages/events/sources/kafka.adoc:
```diff
@@ -0,0 +1,50 @@
+
+= Kafka Consumer (Events)
+
+Events can also be created by routing them to a specific topic on Kafka.
```
You can also create events by routing them to a specific topic on Kafka.

Bonrob2 replied:
```suggestion
You can also create events by routing them to a specific topic on Kafka.
```

Bonrob2 commented on docs/modules/operation/pages/events/sources/kafka.adoc:
```diff
@@ -0,0 +1,50 @@
+= Kafka Consumer (Events)
+
+Events can also be created by routing them to a specific topic on Kafka.
+The Kafka Consumer feature allows events to be consumed from specific topic on Kafka.
```
The Kafka Consumer feature allows events to be consumed from a specific topic on Kafka.

Bonrob2 replied:
```suggestion
The Kafka Consumer feature allows events to be consumed from a specific topic on Kafka.
```

Bonrob2 commented on docs/modules/operation/pages/events/sources/kafka.adoc:
```diff
@@ -0,0 +1,50 @@
+Events can also be created by routing them to a specific topic on Kafka.
+The Kafka Consumer feature allows events to be consumed from specific topic on Kafka.
+
+Events should be created in protobuf format link:https://developers.google.com/protocol-buffers/[Google Protocol Buffers (GPB)]
```
Create events in protobuf fomrat link:https://developers.google.com/protocol-buffers/[Google Protocol Buffers (GPB)].

Bonrob2 commented on docs/modules/operation/pages/events/sources/kafka.adoc:
```diff
@@ -0,0 +1,50 @@
+Events should be created in protobuf format link:https://developers.google.com/protocol-buffers/[Google Protocol Buffers (GPB)]
+See `kafka-consumer-events.proto`  for the model definitions.
+
+The Kafka Consumer by default consumes events from topic named `opennms-events` and forward them to `Eventd`.
```
The Kafka Consumer by default consumes events from topics named `opennms-events` and forwards them to `Eventd`.

Bonrob2 replied:
```suggestion
The Kafka Consumer by default consumes events from topics named `opennms-events` and forwards them to `Eventd`.
```

Bonrob2 commented on docs/modules/operation/pages/events/sources/kafka.adoc:
```diff
@@ -0,0 +1,50 @@
+Events can also be created by routing them to a specific topic on Kafka.
+The Kafka Consumer feature allows events to be consumed from specific topic on Kafka.
+
+Events should be created in protobuf format link:https://developers.google.com/protocol-buffers/[Google Protocol Buffers (GPB)]
```
Events should be created in protobuf format

Bonrob2 replied:
```suggestion
Create events in protobuf format link:https://developers.google.com/protocol-buffers/[Google Protocol Buffers (GPB)].
```

cgorantla commented:
@Bonrob2 Thanks the review. Updated.

Bonrob2 commented:
LGTM!

j-white reviewed (commented):
Looks good. Much nicer than XML wrapped in a Protobuf payload :)

RangerRick reviewed (approved):
[ship it squirrel]
